### PR TITLE
fix: return error to client if prepare stmt param not match

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

before: (in server)

2023-12-12T11:39:05.369596Z  WARN servers::mysql::server: Internal error occurred during query exec, server actively close the channel to let client try next time err=0: Expected type: Timestamp(Millisecond(TimestampMillisecondType)), actual: MYSQL_TYPE_VAR_STRING, at src/servers/src/mysql/helper.rs:168:14

after: (in client)

com.mysql.cj.jdbc.exceptions.MysqlDataTruncation: Data truncation: err: Expected type: Timestamp(Millisecond(TimestampMillisecondType)), actual: MYSQL_TYPE_VAR_STRING, params: (Value(Bytes([49])), MYSQL_TYPE_VAR_STRING),(Value(Int(1)), MYSQL_TYPE_LONG),(Value(Int(1)), MYSQL_TYPE_LONG),(Value(Int(1)), MYSQL_TYPE_LONG),(Value(Int(1)), MYSQL_TYPE_LONG),(Value(Int(1)), MYSQL_TYPE_LONG),(Value(Int(1)), MYSQL_TYPE_LONG)

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/servers/src/mysql/handler.rs [hunk 1]
```diff
@@ -395,7 +416,7 @@ fn format_duration(duration: Duration) -> String {
 fn replace_params_with_values(
     plan: &LogicalPlan,
     param_types: HashMap<String, Option<ConcreteDataType>>,
-    params: Vec<ParamValue>,
+    params: &Vec<ParamValue>,
```
Comment (waynexia) [inline src/servers/src/mysql/handler.rs:419]: style sugg:
```suggestion
    params: &[ParamValue],
```

File: src/servers/src/mysql/handler.rs [hunk 2]
```diff
@@ -286,7 +287,27 @@ impl<W: AsyncWrite + Send + Sync + Unpin> AsyncMysqlShim<W> for MysqlInstanceShi
                     }
                     .fail();
                 }
-                let plan = replace_params_with_values(&plan, param_types, params)?;
+
+                let plan = match replace_params_with_values(&plan, param_types, &params) {
+                    Ok(plan) => plan,
+                    Err(e) => {
+                        w.error(
+                            ErrorKind::ER_TRUNCATED_WRONG_VALUE,
```
Comment (waynexia) [inline src/servers/src/mysql/handler.rs:295]: Why shadow this error? What does it mean

**Review by waynexia — APPROVED** [review]
Rest LGTM

File: src/servers/src/mysql/handler.rs (hunk 2 above)
Comment (MichaelScofield) [reply src/servers/src/mysql/handler.rs:295]: It's not shadow this error. It's returning to client. It means the parameter values and types don't match in the prepare stmt.

**Review by killme2008 — APPROVED** [review]
LGTM. But can we add a test to https://github.com/GreptimeTeam/greptimedb/blob/develop/tests-integration/tests/sql.rs

**Pull request merged**